core/src/main/scala/org/apache/spark/sql/crossdata/catalog/CatalogChain.scala

```diff
@@ -0,0 +1,134 @@
+
+  private implicit def crossdataTable2tableIdentifier(xdTable: CrossdataTable): TableIdentifier =
+    TableIdentifier(xdTable.tableName, xdTable.dbName)
+
+  val catalogs = (firstLevel +: fallbackCatalogs)
+
+  private def chainedLookup[R](lookup: XDCatalog => Option[R]): Option[R] =
+    catalogs.view map(lookup) collectFirst {
+      case Some(res) => res
+    }
+
+  private def persistentCatalogsView: SeqView[PersistentCatalog, Seq[_]] = catalogs.view collect {
+    case c: PersistentCatalog => c
+  }
+
+  // Persistence interface
+
+  override def persistTable(crossdataTable: CrossdataTable, table: LogicalPlan): Unit = {
+    registerTable(crossdataTable, table)
+    persistentCatalogsView foreach (_.persistTable(crossdataTable, table))
+  }
+
+  override def persistView(viewIdentifier: ViewIdentifier, plan: LogicalPlan, sqlText: String): Unit = {
+    registerView(viewIdentifier, plan)
+    persistentCatalogsView foreach (_.persistView(viewIdentifier, plan, sqlText))
+  }
+
+  override def dropTable(tableIdentifier: TableIdentifier): Unit = {
+    if (!tableExists(tableIdentifier)) throw new RuntimeException("Table can't be deleted because it doesn't exists")
+    logInfo(s"Deleting table ${tableIdentifier.unquotedString}from catalog")
+    unregisterTable(tableIdentifier)
+    persistentCatalogsView foreach (_.dropTable(tableIdentifier))
+  }
+
+  override def dropView(viewIdentifier: ViewIdentifier): Unit = {
+    if (!tableExists(viewIdentifier)) throw new RuntimeException("View can't be deleted because it doesn't exists")
+    logInfo(s"Deleting table ${viewIdentifier.unquotedString} from catalog")
+    unregisterView(viewIdentifier)
+    persistentCatalogsView foreach (_.dropView(viewIdentifier))
+  }
+
+  override def dropAllTables(): Unit = {
+    dropAllViews
+    unregisterAllTables
+    persistentCatalogsView foreach (_.dropAllTables)
+  }
+
+  override def dropAllViews(): Unit = {
+    unregisterAllViews
+    persistentCatalogsView foreach (_.dropAllViews)
+  }
+
+  // Catalog interface
+
+  override def registerTable(tableIdent: TableIdentifier, plan: LogicalPlan): Unit =
+    catalogs foreach (_.registerTable(tableIdent, plan))
+  override def registerView(viewIdentifier: ViewIdentifier, plan: LogicalPlan): Unit =
+    catalogs foreach (_.registerView(viewIdentifier, plan))
+
+  override def unregisterTable(tableIdent: TableIdentifier): Unit = catalogs foreach (_.unregisterTable(tableIdent))
+  override def unregisterView(viewIdent: ViewIdentifier): Unit = catalogs foreach (_.unregisterView(viewIdent))
+
+  override def unregisterAllTables(): Unit = catalogs foreach(_.unregisterAllTables)
+  override def unregisterAllViews(): Unit = catalogs.foreach(_.unregisterAllViews)
+
+  private def lookupRelationOpt(tableIdent: TableIdentifier, alias: Option[String]): Option[LogicalPlan] =
+    chainedLookup { catalog: XDCatalog =>
+      if(catalog tableExists tableIdent) Some(catalog.lookupRelation(tableIdent, alias)) else None
```

Comment:
Flow control through try-catch is usually regarded as an anti-pattern
http://programmers.stackexchange.com/questions/189222/are-exceptions-as-control-flow-considered-a-serious-antipattern-if-so-why